contracts/halo-pool/src/contract.rs

```diff
@@ -139,114 +196,320 @@ pub fn execute_add_reward_balance(
+    // Get current pool index
+    let current_phase_index = pool_infos.current_phase_index;
+    // Init response
+    let mut res = Response::new();
+
+    // Not allow removing activated phase
+    if phase_index <= current_phase_index {
+        return Err(ContractError::Std(StdError::generic_err(
+            "Can not remove activated phase",
+        )));
+    }
+
+    // Check the message sender is the whitelisted address
+    if !pool_infos.whitelist.contains(&info.sender) {
+        return Err(ContractError::Std(StdError::generic_err(
+            "Unauthorized: Sender is not in the whitelisted address",
+        )));
+    }
+
+    // If the pool already added reward balance, transfer back all phase reward balance to the sender
+    if pool_infos.reward_balance[phase_index as usize] > Uint128::zero() {
+        let transfer_reward = match pool_infos.reward_token.clone() {
+            TokenInfo::Token { contract_addr } => SubMsg::new(CosmosMsg::Wasm(WasmMsg::Execute {
+                contract_addr,
+                msg: to_binary(&Cw20ExecuteMsg::Transfer {
+                    recipient: info.sender.to_string(),
+                    amount: pool_infos.reward_balance[phase_index as usize],
+                })?,
+                funds: vec![],
+            })),
+            TokenInfo::NativeToken { denom } => SubMsg::new(CosmosMsg::Bank(BankMsg::Send {
+                to_address: info.sender.to_string(),
+                amount: vec![coin(
+                    pool_infos.reward_balance[phase_index as usize].into(),
+                    denom,
+                )],
+            })),
+        };
+        res = res.add_submessage(transfer_reward);
+    }
+
+    // Remove phase
+    pool_infos.pool_infos.remove(phase_index as usize);
+    pool_infos.reward_balance.remove(phase_index as usize);
+    pool_infos.last_reward_time.remove(phase_index as usize);
+    pool_infos
+        .accrued_token_per_share
+        .remove(phase_index as usize);
+
+    // Save pool infos
+    POOL_INFOS.save(deps.storage, &pool_infos)?;
+
+    Ok(res.add_attribute("method", "remove_phase"))
+}
+
 pub fn execute_deposit(
     deps: DepsMut,
     env: Env,
     info: MessageInfo,
     amount: Uint128,
 ) -> Result<Response, ContractError> {
     // Get current time
-    let current_time = env.block.time;
-    // Get pool info
-    let pool_info: PoolInfo = POOL_INFO.load(deps.storage)?;
+    let current_time = env.block.time.seconds();
+    // Get pool infos
+    let mut pool_infos = POOL_INFOS.load(deps.storage)?;
+    // Get current pool index
+    let current_phase_index = pool_infos.current_phase_index;
+    // Get current pool info in pool infos
+    let pool_info = pool_infos.pool_infos[current_phase_index as usize].clone();
     // get staker info
     let mut staker_info = STAKERS_INFO
         .may_load(deps.storage, info.sender.clone())?
-        .unwrap_or(StakerRewardAssetInfo {
+        .unwrap_or(StakerInfoResponse {
```

Comment:
chỗ này có thể lỗi vì "may_load will parse the data stored at the key if present, returns Ok(None) if no data there.". Nếu không có data nó vẫn trả về Ok nên cái unwrap_or sợ là có thể bị lỗi. E đổi lại thành check điều kiện nếu có data và init nếu chưa có data riêng cho an toàn

Reply:
Updated as your comment